Mathlib/RingTheory/Polynomial/SmallDegreeVieta.lean

```diff
@@ -100,4 +111,58 @@ lemma aroots_quadratic_eq_pair_iff_of_ne_zero' [CommRing T] [Field S] [Algebra T
     X ^ 2 + C ((algebraMap T S) b) * X + C ((algebraMap T S) c) by simp]
   exact roots_quadratic_eq_pair_iff_of_ne_zero' ha
 
+/-
+Polynomial versions of results in `Algebra.QuadraticDiscriminant`
+-/
+section QuadraticDiscriminant
+
+variable [Field R] {a b c : R}
+
+/-- Roots of a quadratic equation. -/
+theorem isRoot_quadratic_iff [NeZero (2 : R)] (ha : a ≠ 0) {s : R}
+    (h : discrim a b c = s * s) (x : R) :
+    IsRoot (C a * X ^ 2 + C b * X + C c) x ↔ x = (-b + s) / (2 * a) ∨ x = (-b - s) / (2 * a) := by
+  rw [← quadratic_eq_zero_iff ha h]; simp [pow_two]
+
+/-- Root of a quadratic when its discriminant equals zero -/
+theorem isRoot_quadratic_iff_of_discrim_eq_zero [NeZero (2 : R)] (ha : a ≠ 0)
+    (h : discrim a b c = 0) (x : R) :
+    IsRoot (C a * X ^ 2 + C b * X + C c) x ↔ x = -b / (2 * a) := by
+  rw [← quadratic_eq_zero_iff_of_discrim_eq_zero ha h]; simp [pow_two]
+
+/-- A quadratic has no root if its discriminant has no square root. -/
+theorem not_isRoot_of_discrim_ne_sq (h : ∀ s : R, discrim a b c ≠ s^2) (x : R) :
+    ¬ IsRoot (C a * X ^ 2 + C b * X + C c) x := by
+  convert quadratic_ne_zero_of_discrim_ne_sq h x using 1; simp [pow_two]
+
+theorem mem_roots_quadratic_iff_of_discrim_eq_sq [NeZero (2 : R)] (ha : a ≠ 0)
+    {z s : R} (h : discrim a b c = s * s) :
+    z ∈ (C a * X ^ 2 + C b * X + C c).roots ↔ z = (-b + s) / (2 * a) ∨ z = (-b - s) / (2 * a) := by
+  rw [mem_roots (quadratic_ne_zero ha), isRoot_quadratic_iff ha h]
+
+theorem vieta_of_discrim_eq_sq [NeZero (2 : R)] (ha : a ≠ 0) {s : R} (h : discrim a b c = s * s) :
+    b = -a * ((-b + s) / (2 * a) + (-b - s) / (2 * a)) ∧
+      c = a * ((-b + s) / (2 * a)) * ((-b - s) / (2 * a)) := by
+  ring_nf
+  rw [sq s, ← h, discrim]
+  field_simp
+  ring_nf
+
+theorem quadratic_eq_of_discrim_eq_sq [NeZero (2 : R)] (ha : a ≠ 0) {s : R}
+    (h : discrim a b c = s * s) : C a * X ^ 2 + C b * X + C c =
+      C a * (X - C ((-b + s) / (2 * a))) * (X - C ((-b - s) / (2 * a))) :=
+  quadratic_eq_of_vieta (vieta_of_discrim_eq_sq ha h)
+
+theorem roots_quadratic_of_discrim_ne_sq (h : ∀ s : R, discrim a b c ≠ s^2) :
```

Comment:
You're using a mixture of `s * s` and `s ^ 2`: this will make for a very annoying API. I think either is fine but we must be consistent.

Reply:
Okay, let's use `s ^ 2`.